File: 100-999/101. Symmetric Tree/solution_2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public bool WalkSymmetric(TreeNode left, TreeNode right)
    {
        if (left == null && right == null)
            return true;
        if (left == null || right == null)
            return false;

        return WalkSymmetric(left.left, right.right) && WalkSymmetric(left.right, right.left) && left.val == right.val;
    }
    
    public bool IsSymmetric(TreeNode root)
    {
        if (root != null)
            return WalkSymmetric(root.left, root.right);

        return false;
    }
}

File: 100-999/101. Symmetric Tree/solution_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution 
{
    public:
        bool walkSymmetric(TreeNode* left, TreeNode* right)
        {
            if (!left && !right)
                return true;
            
            if (!left || !right)
                return false;

            
            return walkSymmetric(left->left, right->right) && walkSymmetric(left->right, right->left) && left->val == right->val;        
        }

        bool isSymmetric(TreeNode* root) 
        {
            if (root)
                return walkSymmetric(root->left, root->right);
  
            return false;
        }
};